File: frontend/src/api/budget.ts
import api from './client'
import type { ExpenseCategoryId } from '@/types/budget'

export interface BudgetTagResponse {
  id: number
  name: string
  color: string
}

export interface BudgetTagCreate {
  name: string
  color?: string
}

export interface BudgetTagUpdate {
  name?: string
  color?: string
}

export interface BudgetCategoryResponse {
  id: number
  key: string
  label: string
  icon: string
  color: string
  sort_order: number
  is_archived: boolean
}

export interface BudgetCategoryCreate {
  key?: string
  label: string
  icon?: string
  color?: string
  sort_order?: number
}

export interface BudgetCategoryUpdate {
  label?: string
  icon?: string
  color?: string
  sort_order?: number
  is_archived?: boolean
}

export interface TransactionResponse {
  id: number
  type: 'expense' | 'income'
  amount: number
  category: ExpenseCategoryId | null
  description: string
  date: string
  created_at: string
  tags: BudgetTagResponse[]
  source_planned_id?: number | null
}

export interface TransactionCreate {
  type: 'expense' | 'income'
  amount: number
  category: ExpenseCategoryId | null
  description: string
  date: string
  tag_ids?: number[]
}

export interface TransactionUpdate {
  type?: 'expense' | 'income'
  amount?: number
  category?: ExpenseCategoryId | null
  description?: string
  date?: string
  tag_ids?: number[]
}

export interface PlannedPurchaseResponse {
  id: number
  year: number
  month: number
  amount: number
  category: ExpenseCategoryId | null
  description: string
  done: boolean
  created_at: string
  tags: BudgetTagResponse[]
  source_recurring_id?: number | null
}

export interface PlannedPurchaseCreate {
  year: number
  month: number
  amount: number
  category: ExpenseCategoryId | null
  description: string
  done?: boolean
  tag_ids?: number[]
}

export interface PlannedPurchaseUpdate {
  amount?: number
  category?: ExpenseCategoryId | null
  description?: string
  done?: boolean
  tag_ids?: number[]
}

export interface PlannedConvertRequest {
  amount?: number
  date?: string
  tag_ids?: number[]
}

export interface AllocationResponse {
  id: number
  year: number
  month: number
  category: string
  limit_amount: number
  created_at: string
}

export interface AllocationUpsert {
  year: number
  month: number
  category: string
  limit_amount: number
}

export interface RecurringResponse {
  id: number
  type: 'expense' | 'income'
  amount: number
  category: string | null
  description: string
  tag_ids: number[]
  day_of_month: number
  start_date: string
  end_date: string | null
  last_generated_date: string | null
  is_paused: boolean
  created_at: string
}

export interface RecurringCreate {
  type: 'expense' | 'income'
  amount: number
  category?: string | null
  description?: string
  tag_ids?: number[]
  day_of_month: number
  start_date: string
  end_date?: string | null
  is_paused?: boolean
}

export interface RecurringUpdate {
  type?: 'expense' | 'income'
  amount?: number
  category?: string | null
  description?: string
  tag_ids?: number[]
  day_of_month?: number
  start_date?: string
  end_date?: string | null
  is_paused?: boolean
}

// ── Summary types ─────────────────────────────────────────────────────────────

export interface SummaryTotals {
  income: number
  expense: number
  balance: number
  planned_pending: number
  planned_done: number
}

export interface SummaryCategory {
  category: string | null
  allocated: number
  spent: number
  planned_pending: number
  remaining: number
  pct: number
}

export interface SummaryDay {
  date: string
  expense: number
  income: number
}

export interface SummaryTrend {
  prev_expense: number
  delta_abs: number
  delta_pct: number | null
}

export interface SummaryProjection {
  days_passed: number
  days_total: number
  days_left: number
  rate_per_day: number
  projected_total: number
  daily_budget: number
}

export interface SummaryResponse {
  year: number
  month: number
  totals: SummaryTotals
  by_category: SummaryCategory[]
  daily: SummaryDay[]
  trend: SummaryTrend
  projection: SummaryProjection
}

// ── History types ─────────────────────────────────────────────────────────────

export interface HistoryQuery {
  q?: string
  type?: 'expense' | 'income'
  tag_ids?: number[]
  category?: string
  from?: string
  to?: string
  amount_min?: number
  amount_max?: number
  sort?: 'date' | 'amount'
  order?: 'asc' | 'desc'
  limit?: number
  offset?: number
}

export interface HistoryResponse {
  items: TransactionResponse[]
  total: number
  offset: number
  limit: number
}

export const budgetApi = {
  // Tags
  listTags: () =>
    api.get<BudgetTagResponse[]>('/budget/tags').then((r) => r.data),

  createTag: (data: BudgetTagCreate) =>
    api.post<BudgetTagResponse>('/budget/tags', data).then((r) => r.data),

  updateTag: (id: number, data: BudgetTagUpdate) =>
    api.put<BudgetTagResponse>(`/budget/tags/${id}`, data).then((r) => r.data),

  deleteTag: (id: number) =>
    api.delete(`/budget/tags/${id}`),

  // Categories
  listCategories: (includeArchived = false) =>
    api
      .get<BudgetCategoryResponse[]>('/budget/categories', { params: { include_archived: includeArchived } })
      .then((r) => r.data),

  createCategory: (data: BudgetCategoryCreate) =>
    api.post<BudgetCategoryResponse>('/budget/categories', data).then((r) => r.data),

  updateCategory: (id: number, data: BudgetCategoryUpdate) =>
    api.put<BudgetCategoryResponse>(`/budget/categories/${id}`, data).then((r) => r.data),

  deleteCategory: (id: number) =>
    api.delete(`/budget/categories/${id}`),

  // Allocations
  listAllocations: (year: number, month: number) =>
    api.get<AllocationResponse[]>('/budget/allocations', { params: { year, month } }).then((r) => r.data),

  upsertAllocation: (data: AllocationUpsert) =>
    api.post<AllocationResponse>('/budget/allocations', data).then((r) => r.data),

  deleteAllocation: (id: number) =>
    api.delete(`/budget/allocations/${id}`),

  // Transactions
  listTransactions: (year?: number, month?: number) =>
    api.get<TransactionResponse[]>('/budget/transactions', { params: { year, month } }).then((r) => r.data),

  createTransaction: (data: TransactionCreate) =>
    api.post<TransactionResponse>('/budget/transactions', data).then((r) => r.data),

  updateTransaction: (id: number, data: TransactionUpdate) =>
    api.put<TransactionResponse>(`/budget/transactions/${id}`, data).then((r) => r.data),

  deleteTransaction: (id: number) =>
    api.delete(`/budget/transactions/${id}`),

  // Planned
  listPlanned: (year?: number, month?: number) =>
    api.get<PlannedPurchaseResponse[]>('/budget/planned', { params: { year, month } }).then((r) => r.data),

  createPlanned: (data: PlannedPurchaseCreate) =>
    api.post<PlannedPurchaseResponse>('/budget/planned', data).then((r) => r.data),

  updatePlanned: (id: number, data: PlannedPurchaseUpdate) =>
    api.put<PlannedPurchaseResponse>(`/budget/planned/${id}`, data).then((r) => r.data),

  deletePlanned: (id: number) =>
    api.delete(`/budget/planned/${id}`),

  convertPlanned: (id: number, data: PlannedConvertRequest) =>
    api.post<TransactionResponse>(`/budget/planned/${id}/convert`, data).then((r) => r.data),

  // Recurring
  listRecurring: () =>
    api.get<RecurringResponse[]>('/budget/recurring').then((r) => r.data),

  createRecurring: (data: RecurringCreate) =>
    api.post<RecurringResponse>('/budget/recurring', data).then((r) => r.data),

  updateRecurring: (id: number, data: RecurringUpdate) =>
    api.put<RecurringResponse>(`/budget/recurring/${id}`, data).then((r) => r.data),

  deleteRecurring: (id: number) =>
    api.delete(`/budget/recurring/${id}`),

  // Summary / History
  getSummary: (year: number, month: number) =>
    api.get<SummaryResponse>('/budget/summary', { params: { year, month } }).then((r) => r.data),

  getHistory: (q: HistoryQuery = {}) =>
    api.get<HistoryResponse>('/budget/history', { params: q }).then((r) => r.data),

  exportUrl: (q: { year?: number; month?: number; from?: string; to?: string; type?: 'expense' | 'income' } = {}) => {
    const params = new URLSearchParams()
    Object.entries(q).forEach(([k, v]) => {
      if (v !== undefined && v !== null) params.set(k, String(v))
    })
    const qs = params.toString()
    return `/api/budget/export${qs ? '?' + qs : ''}`
  },

  downloadCsv: async (q: { year?: number; month?: number; from?: string; to?: string; type?: 'expense' | 'income' } = {}) => {
    const response = await api.get('/budget/export', { params: q, responseType: 'blob' })
    const blob = new Blob([response.data], { type: 'text/csv;charset=utf-8' })
    const url = window.URL.createObjectURL(blob)
    const link = document.createElement('a')
    link.href = url
    const disposition = response.headers['content-disposition'] as string | undefined
    const match = disposition && /filename="?([^"]+)"?/.exec(disposition)
    link.download = match ? match[1] : 'budget.csv'
    document.body.appendChild(link)
    link.click()
    link.remove()
    window.URL.revokeObjectURL(url)
  },
}

File: backend/app/schemas/budget.py
from datetime import datetime
from typing import Literal

from pydantic import BaseModel


# ── Budget Tags ────────────────────────────────────────────────────────────────

class BudgetTagCreate(BaseModel):
    name: str
    color: str = "#6B7280"


class BudgetTagUpdate(BaseModel):
    name: str | None = None
    color: str | None = None


class BudgetTagResponse(BaseModel):
    id: int
    name: str
    color: str

    model_config = {"from_attributes": True}


# ── Budget Categories ──────────────────────────────────────────────────────────

class BudgetCategoryCreate(BaseModel):
    key: str | None = None  # если пусто — сгенерим из label
    label: str
    icon: str = "📦"
    color: str = "#9CA3AF"
    sort_order: int | None = None


class BudgetCategoryUpdate(BaseModel):
    label: str | None = None
    icon: str | None = None
    color: str | None = None
    sort_order: int | None = None
    is_archived: bool | None = None


class BudgetCategoryResponse(BaseModel):
    id: int
    key: str
    label: str
    icon: str
    color: str
    sort_order: int
    is_archived: bool

    model_config = {"from_attributes": True}


# ── Transactions ───────────────────────────────────────────────────────────────

class TransactionCreate(BaseModel):
    type: Literal["expense", "income"]
    amount: float
    category: str | None = None
    description: str = ""
    date: str  # yyyy-MM-dd
    tag_ids: list[int] = []


class TransactionUpdate(BaseModel):
    type: Literal["expense", "income"] | None = None
    amount: float | None = None
    category: str | None = None
    description: str | None = None
    date: str | None = None
    tag_ids: list[int] | None = None


class TransactionResponse(BaseModel):
    id: int
    type: str
    amount: float
    category: str | None
    description: str
    date: str
    created_at: datetime
    tags: list[BudgetTagResponse] = []
    source_planned_id: int | None = None

    model_config = {"from_attributes": True}


# ── Planned Purchases ──────────────────────────────────────────────────────────

class PlannedPurchaseCreate(BaseModel):
    year: int
    month: int  # 0-based
    amount: float
    category: str | None = None
    description: str = ""
    done: bool = False
    tag_ids: list[int] = []


class PlannedPurchaseUpdate(BaseModel):
    amount: float | None = None
    category: str | None = None
    description: str | None = None
    done: bool | None = None
    tag_ids: list[int] | None = None


class PlannedPurchaseResponse(BaseModel):
    id: int
    year: int
    month: int
    amount: float
    category: str | None
    description: str
    done: bool
    created_at: datetime
    tags: list[BudgetTagResponse] = []
    source_recurring_id: int | None = None

    model_config = {"from_attributes": True}


class PlannedConvertRequest(BaseModel):
    amount: float | None = None
    date: str | None = None  # yyyy-MM-dd, default = today
    tag_ids: list[int] | None = None


# ── Budget Allocations ─────────────────────────────────────────────────────────

class AllocationUpsert(BaseModel):
    year: int
    month: int  # 0-based
    category: str
    limit_amount: float


class AllocationResponse(BaseModel):
    id: int
    year: int
    month: int
    category: str
    limit_amount: float
    created_at: datetime

    model_config = {"from_attributes": True}


# ── Recurring Transactions ─────────────────────────────────────────────────────

class RecurringCreate(BaseModel):
    type: Literal["expense", "income"]
    amount: float
    category: str | None = None
    description: str = ""
    tag_ids: list[int] = []
    day_of_month: int
    start_date: str  # yyyy-MM-dd
    end_date: str | None = None
    is_paused: bool = False


class RecurringUpdate(BaseModel):
    type: Literal["expense", "income"] | None = None
    amount: float | None = None
    category: str | None = None
    description: str | None = None
    tag_ids: list[int] | None = None
    day_of_month: int | None = None
    start_date: str | None = None
    end_date: str | None = None
    is_paused: bool | None = None


class RecurringResponse(BaseModel):
    id: int
    type: str
    amount: float
    category: str | None
    description: str
    tag_ids: list[int]
    day_of_month: int
    start_date: str
    end_date: str | None
    last_generated_date: str | None
    is_paused: bool
    created_at: datetime

    model_config = {"from_attributes": True}


# ── Summary (dashboard aggregate) ──────────────────────────────────────────────

class SummaryTotals(BaseModel):
    income: float
    expense: float
    balance: float
    planned_pending: float
    planned_done: float


class SummaryCategory(BaseModel):
    category: str | None
    allocated: float  # 0 if no allocation
    spent: float
    planned_pending: float
    remaining: float  # allocated - spent - planned_pending
    pct: float  # spent / allocated * 100, 0 if no allocation


class SummaryDay(BaseModel):
    date: str  # yyyy-MM-dd
    expense: float
    income: float


class SummaryTrend(BaseModel):
    prev_expense: float
    delta_abs: float
    delta_pct: float | None  # None when prev is 0


class SummaryProjection(BaseModel):
    days_passed: int
    days_total: int
    days_left: int
    rate_per_day: float
    projected_total: float
    daily_budget: float  # sum of allocations / days_total


class SummaryResponse(BaseModel):
    year: int
    month: int
    totals: SummaryTotals
    by_category: list[SummaryCategory]
    daily: list[SummaryDay]
    trend: SummaryTrend
    projection: SummaryProjection


# ── History (filtered paginated list) ──────────────────────────────────────────

class HistoryResponse(BaseModel):
    items: list[TransactionResponse]
    total: int
    offset: int
    limit: int
